src/models/tests/pairwise_union_reach_surface_test.py

```diff
@@ -0,0 +1,118 @@
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+
+class LinearCappedReachCurve(ReachCurve):
+  """Linear ReachCurve that is capped at max_reach."""
+
+  def _fit(self) -> None:
+    self._max_reach = self._data[0].reach()
+
+  def by_impressions(self,
+                     impressions: [int],
+                     max_frequency: int = 1) -> ReachPoint:
+    """Returns the estimated reach for a given impression vector."""
+    kplus_frequencies = [
+        min(sum(impressions), self.max_reach) // i
+        for i in range(1, max_frequency + 1)
+    ]
+    return ReachPoint(impressions, kplus_frequencies)
+
+
+class PairwiseUnionReachSurfaceTest(absltest.TestCase):
+
+  def assertPointsAlmostEqualToPrediction(self,
+                                          surface,
+                                          reach_points,
+                                          tolerance=0.0001):
+    for reach_point in reach_points:
+      self.assertAlmostEqual(
+          surface.by_impressions(reach_point.impressions).reach(),
+          reach_point.reach(),
+          delta=reach_point.reach() * tolerance)
+
+  def generate_true_reach(self, a, reach_curves, impressions):
+    p = len(reach_curves)
+    reach_vector = [
+        reach_curve.by_impressions(impression).reach()
+        for reach_curve, impression in zip(reach_curves, impressions)
+    ]
+    reach = sum(reach_vector)
+    for i in range(len(reach_curves)):
+      for j in range(len(reach_curves)):
+        reach -= (a[i * p + j] * reach_vector[i] * reach_vector[j]) / (
+            max(reach_curves[i].max_reach, reach_curves[j].max_reach) * 2)
+    return ReachPoint(impressions, [reach])
+
+  def generate_sample_reach_curves(self, num_publishers, decay_rate,
+                                   universe_size):
+    max_reaches = [
+        universe_size * (decay_rate**pub_num)
+        for pub_num in range(num_publishers)
+    ]
+    return [
+        LinearCappedReachCurve([ReachPoint([max_reach], (max_reach,))])
+        for max_reach in max_reaches
+    ]
+
+  def generate_sample_matrix_a(self, num_publishers):
+    true_a = np.array([
+        np.random.dirichlet(np.ones(num_publishers)) * np.random.uniform()
+        for _ in range(num_publishers)
+    ]).flatten()
+    for i in range(num_publishers):
+      true_a[i * num_publishers + i] = 0
+    return true_a
+
+  def generate_sample_reach_points(self, true_a, reach_curves, size,
+                                   universe_size):
+    reach_points = []
+    random_generator = np.random.default_rng(1)
+    for _ in range(size):
+      impressions = [[random_generator.uniform(0, universe_size / 2)]
+                     for _ in range(len(reach_curves))]
+      reach_points.append(
+          self.generate_true_reach(true_a, reach_curves, impressions))
+
+    return reach_points
+
+  def test_by_impressions(self):
```

Comment:
When I run this unit test, I get some additional output, as follows:

```
     pcost       dcost       gap    pres   dres
 0: -3.4733e+09 -3.4733e+09  2e+01  2e+00  3e-10
 1: -3.4733e+09 -3.4733e+09  3e+00  2e-02  3e-12
 2: -3.4733e+09 -3.4733e+09  3e-02  2e-04  3e-14
 3: -3.4733e+09 -3.4733e+09  3e-04  2e-06  3e-16
 4: -3.4733e+09 -3.4733e+09  4e-06  2e-08  6e-17
Optimal solution found.
```

Can you turn this off?

Reply:
Try add `solvers.options['show_progress'] = False` in the beginning

Reply:
thank you! I added at the beginning of the pairwise_union_reach_surface file not the test. What do you think? Do you prefer I add to test or is this good?

Reply:
It's better to have it in pairwise_union_reach_surface.py (as it is now).  Thanks!